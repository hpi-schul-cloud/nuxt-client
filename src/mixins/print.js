import kjua from "kjua";

const printStyles = (optionalStyles) => `<style>
	@page {
		size: A4;
		margin: 16px;
	}
	body {
		font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Ubuntu, Helvetica Neue, Arial, sans-serif;
	}
	${optionalStyles}
</style>`;

const getPageHtml = (head, body) => `<html>
	<head>${head}</head>
	<body>${body}</body>
</html>`;

const getQRCodeBase64Image = (text) =>
	kjua({ text: text, render: "canvas" }).toDataURL();

export default {
	methods: {
		$_print(content, styles = "") {
			const w = window.open();
			w.document.write(getPageHtml(printStyles(styles), content));
			w.document.close();
			/* eventListener is needed to give the browser some rendering time for the image */
			w.addEventListener("load", () => {
				w.focus();
				w.print();
				// why is this timeout required?
				setTimeout(function () {
					w.close();
				}, 500);
			});
		},

		$_printQRs(items = []) {
			const styles = `
				.part{
					border: 1px solid #aaa;
					width: 4cm;
					float: left;
					padding: 8px;
					margin: 4px;
				}
				.qr-code{ width: 100% !important; height: auto !important; }
				.qr-content{
					font-size: 0.7em;
					opacity: 0.7;
					width: 100%;
					word-break:
					break-all;
					word-break: break-word;
					margin: 2px 0 0;
				}
				.title{ margin: 4px 0; font-size: 1.25em; font-weight: bold; }
				.description{
					font-size: 1em;
					color: #555;
					margin: 0;
					word-break: break-all;
					word-break: break-word;
				}`;
			let content = "";
			if (items.length === 0) {
				content = "Keine EintrÃ¤ge zu drucken.";
			} else {
				content = items
					.map((item) => {
						const QRCodeBase64Image = getQRCodeBase64Image(item.qrContent);
						return `<div class="part">
							<div class="image-wrapper">
								<img class="qr-code" alt="${
									item.qrContent
								}" src="${QRCodeBase64Image}" width="200" height="200">
								<div class="qr-content">${item.qrContent}</div>
							</div>
							${item.title ? `<h4 class="title">${item.title}</h4>` : ""}
							${item.description ? `<p class="description">${item.description}</p>` : ""}
						</div>`;
					})
					.join("\n");
			}
			return this.$_print(content, styles);
		},
	},
};
