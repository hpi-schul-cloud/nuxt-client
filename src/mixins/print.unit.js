import print from "./print";

const getNewWindowMock = () => {
	let content = "";
	return {
		document: {
			get innerHTML() {
				return content;
			},
			write: vi.fn().mockImplementation((...args) => {
				args.forEach((text) => (content += text));
			}),
			close: vi.fn(),
		},
		addEventListener: vi.fn().mockImplementation((event, cb) => cb()),
		focus: vi.fn(),
		print: vi.fn(),
		close: vi.fn(),
	};
};

let newWindowMock;

describe("@/mixins/print", () => {
	beforeEach(() => {
		newWindowMock = getNewWindowMock();
		vi.spyOn(window, "open")
			.mockImplementation()
			.mockReturnValue(newWindowMock);
	});

	describe("$_print", () => {
		const method = print.methods.$_print;

		it("can print plain content", () => {
			const testContent = "some plain old content";
			method(testContent);
			expect(newWindowMock.document.innerHTML).toContain(testContent);
		});

		it("can print content with custom styles", () => {
			const testContent = "some plain old content";
			const testStyles = "body { margin: 2rem; border: 1px solid blue; }";
			method(testContent, testStyles);
			expect(newWindowMock.document.innerHTML).toContain(testContent);
		});

		it("focuses new window before printing", () => {
			const testContent = "some plain old content";
			method(testContent);

			expect(newWindowMock.document.write).toHaveBeenCalledWith(
				expect.stringContaining(testContent)
			);
			expect(newWindowMock.focus).toHaveBeenCalled();
			expect(newWindowMock.print).toHaveBeenCalled();
		});

		it("closes new window after print", () => {
			const testContent = "some plain old content";
			vi.useFakeTimers();
			method(testContent);
			vi.runAllTimers();
			expect(newWindowMock.print).toHaveBeenCalled();
			expect(newWindowMock.close).toHaveBeenCalled();
		});
	});

	describe("$_printQRs", () => {
		const method = print.methods.$_printQRs;

		it("can print items with all options", () => {
			const testContent = [
				{
					qrContent: "qrContent",
					title: "title",
					description: "description",
				},
			];
			method(testContent);
			expect(newWindowMock.document.innerHTML).toContain(
				testContent[0].qrContent
			);
			expect(newWindowMock.document.innerHTML).toContain(testContent[0].title);
			expect(newWindowMock.document.innerHTML).toContain(
				testContent[0].description
			);
		});

		it("can print items with only QR content", () => {
			const testContent = [
				{
					qrContent: "qrContent",
				},
			];
			method(testContent);
			expect(newWindowMock.document.innerHTML).toContain(
				testContent[0].qrContent
			);
		});

		it("prints an error if no items to print are given", () => {
			method([]);
			expect(newWindowMock.document.innerHTML).toContain(
				"Keine Einträge zu drucken."
			);
		});

		it("can handle no parameters", () => {
			method();
			expect(newWindowMock.document.innerHTML).toContain(
				"Keine Einträge zu drucken."
			);
		});
	});
});
