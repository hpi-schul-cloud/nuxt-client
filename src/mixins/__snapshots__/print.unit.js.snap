// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@/mixins/print $_print can print content with custom styles 1`] = `
<html>

  <head>
    <style>
      @page {
        size: A4;
        margin: 16px;
      }

      body {
        font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Ubuntu, Helvetica Neue, Arial, sans-serif;
      }

      body {
        margin: 2rem;
        border: 1px solid blue;
      }
    </style>
  </head>

  <body>some plain old content</body>

</html>
`;

exports[`@/mixins/print $_print can print plain content 1`] = `
<html>

  <head>
    <style>
      @page {
        size: A4;
        margin: 16px;
      }

      body {
        font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Ubuntu, Helvetica Neue, Arial, sans-serif;
      }
    </style>
  </head>

  <body>some plain old content</body>

</html>
`;

exports[`@/mixins/print $_printQRs can handle no parameters 1`] = `
<html>

  <head>
    <style>
      @page {
        size: A4;
        margin: 16px;
      }

      body {
        font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Ubuntu, Helvetica Neue, Arial, sans-serif;
      }

      .part {
        border: 1px solid #aaa;
        width: 4cm;
        float: left;
        padding: 8px;
        margin: 4px;
      }

      .qr-code {
        width: 100% !important;
        height: auto !important;
      }

      .qr-content {
        font-size: 0.7em;
        opacity: 0.7;
        width: 100%;
        word-break:
          break-all;
        word-break: break-word;
        margin: 2px 0 0;
      }

      .title {
        margin: 4px 0;
        font-size: 1.25em;
        font-weight: bold;
      }

      .description {
        font-size: 1em;
        color: #555;
        margin: 0;
        word-break: break-all;
        word-break: break-word;
      }
    </style>
  </head>

  <body>Keine Einträge zu drucken.</body>

</html>
`;

exports[`@/mixins/print $_printQRs can print items with all options 1`] = `
<html>

  <head>
    <style>
      @page {
        size: A4;
        margin: 16px;
      }

      body {
        font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Ubuntu, Helvetica Neue, Arial, sans-serif;
      }

      .part {
        border: 1px solid #aaa;
        width: 4cm;
        float: left;
        padding: 8px;
        margin: 4px;
      }

      .qr-code {
        width: 100% !important;
        height: auto !important;
      }

      .qr-content {
        font-size: 0.7em;
        opacity: 0.7;
        width: 100%;
        word-break:
          break-all;
        word-break: break-word;
        margin: 2px 0 0;
      }

      .title {
        margin: 4px 0;
        font-size: 1.25em;
        font-weight: bold;
      }

      .description {
        font-size: 1em;
        color: #555;
        margin: 0;
        word-break: break-all;
        word-break: break-word;
      }
    </style>
  </head>

  <body>
    <div class="part">
      <div class="image-wrapper">
        <img class="qr-code" alt="qrContent" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAMgAAADICAYAAACtWK6eAAAABmJLR0QA/wD/AP+gvaeTAAADgklEQVR4nO3dQW7bMBRAwabokXL/E+RO6jZA5VfBoWhKnlkWhUtbfeDig9THtm3bL2DX71cvAFYmEAgCgSAQCH/2/vDz83P2Opbx9fX1379zxd/nrt9rlEe/jx0EgkAgCASCQCAIBIJAIAgEgkAg7A4KjzgyeFrNzEHYqMHcagO+d3vudhAIAoEgEAgCgSAQCAKBIBAIAoHw9KDwCAOsNmrNq333Oz13OwgEgUAQCASBQBAIBIFAEAgEgUA4dVD4zu40LHtndhAIAoEgEAgCgSAQCAKBIBAIAoFgUPiE1QZzo64w5V92EAgCgSAQCAKBIBAIAoEgEAgCgXDqoPCuw6lRpwVnvsdwptXW8xN2EAgCgSAQCAKBIBAIAoEgEAgCgfD0oHDm1ZpXNHMIOHOY+G7P3Q4CQSAQBAJBIBAEAkEgEAQCQSAQPrZt2169iDtabaB2p1N+M9lBIAgEgkAgCASCQCAIBIJAIAgEwu6JwplXax4xc+g2c6C22mnBI2aegpzp0ZrtIBAEAkEgEAQCQSAQBAJBIBAEAmF3ULjagO+KA7WZZn73u16p+ogdBIJAIAgEgkAgCASCQCAIBIJAIDx99ehqpw6PuOuaZw5273Ra8Ag7CASBQBAIBIFAEAgEgUAQCASBQNg9UXjEFYdlVxyorbbmUVY7tfqIHQSCQCAIBIJAIAgEgkAgCASCQCA8PSg8YrWrR0f9W6u568Bx1DP9yZrtIBAEAkEgEAQCQSAQBAJBIBAEAuHUQeEVT6jNtNqar3hS0jsK4YUEAkEgEAQCQSAQBAJBIBAEAmF3ULjaO+ZmuuIVpqOsNrg84uzf0A4CQSAQBAJBIBAEAkEgEAQCQSAQLvGOwlFmXmU5ympD25nrmXnl7CN2EAgCgSAQCAKBIBAIAoEgEAgCgXCJdxQeccXB5SgrDNS+u9OzsINAEAgEgUAQCASBQBAIBIFAEAiEUweFdzXz6tFRnzNzzaO+1wpXvNpBIAgEgkAgCASCQCAIBIJAIAgEgkHhSa54gm/UwHG1z/kJOwgEgUAQCASBQBAIBIFAEAgEgUA4dVB4pysov7vrEHDU54xa8wr/f+wgEAQCQSAQBAJBIBAEAkEgEAQC4elB4cxh2V2tNgQcZYWTgN+5ehROIhAIAoEgEAgCgSAQCAKBIBAIH9u2ba9eBKzKDgJBIBAEAkEgEAQC4S+w0gP+veG1cAAAAABJRU5ErkJggg==" width="200" height="200">
        <div class="qr-content">qrContent</div>
      </div>
      <h4 class="title">title</h4>
      <p class="description">description</p>
    </div>
  </body>

</html>
`;

exports[`@/mixins/print $_printQRs can print items with only QR content 1`] = `
<html>

  <head>
    <style>
      @page {
        size: A4;
        margin: 16px;
      }

      body {
        font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Ubuntu, Helvetica Neue, Arial, sans-serif;
      }

      .part {
        border: 1px solid #aaa;
        width: 4cm;
        float: left;
        padding: 8px;
        margin: 4px;
      }

      .qr-code {
        width: 100% !important;
        height: auto !important;
      }

      .qr-content {
        font-size: 0.7em;
        opacity: 0.7;
        width: 100%;
        word-break:
          break-all;
        word-break: break-word;
        margin: 2px 0 0;
      }

      .title {
        margin: 4px 0;
        font-size: 1.25em;
        font-weight: bold;
      }

      .description {
        font-size: 1em;
        color: #555;
        margin: 0;
        word-break: break-all;
        word-break: break-word;
      }
    </style>
  </head>

  <body>
    <div class="part">
      <div class="image-wrapper">
        <img class="qr-code" alt="qrContent" src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAMgAAADICAYAAACtWK6eAAAABmJLR0QA/wD/AP+gvaeTAAADgklEQVR4nO3dQW7bMBRAwabokXL/E+RO6jZA5VfBoWhKnlkWhUtbfeDig9THtm3bL2DX71cvAFYmEAgCgSAQCH/2/vDz83P2Opbx9fX1379zxd/nrt9rlEe/jx0EgkAgCASCQCAIBIJAIAgEgkAg7A4KjzgyeFrNzEHYqMHcagO+d3vudhAIAoEgEAgCgSAQCAKBIBAIAoHw9KDwCAOsNmrNq333Oz13OwgEgUAQCASBQBAIBIFAEAgEgUA4dVD4zu40LHtndhAIAoEgEAgCgSAQCAKBIBAIAoFgUPiE1QZzo64w5V92EAgCgSAQCAKBIBAIAoEgEAgCgXDqoPCuw6lRpwVnvsdwptXW8xN2EAgCgSAQCAKBIBAIAoEgEAgCgfD0oHDm1ZpXNHMIOHOY+G7P3Q4CQSAQBAJBIBAEAkEgEAQCQSAQPrZt2169iDtabaB2p1N+M9lBIAgEgkAgCASCQCAIBIJAIAgEwu6JwplXax4xc+g2c6C22mnBI2aegpzp0ZrtIBAEAkEgEAQCQSAQBAJBIBAEAmF3ULjagO+KA7WZZn73u16p+ogdBIJAIAgEgkAgCASCQCAIBIJAIDx99ehqpw6PuOuaZw5273Ra8Ag7CASBQBAIBIFAEAgEgUAQCASBQNg9UXjEFYdlVxyorbbmUVY7tfqIHQSCQCAIBIJAIAgEgkAgCASCQCA8PSg8YrWrR0f9W6u568Bx1DP9yZrtIBAEAkEgEAQCQSAQBAJBIBAEAuHUQeEVT6jNtNqar3hS0jsK4YUEAkEgEAQCQSAQBAJBIBAEAmF3ULjaO+ZmuuIVpqOsNrg84uzf0A4CQSAQBAJBIBAEAkEgEAQCQSAQLvGOwlFmXmU5ympD25nrmXnl7CN2EAgCgSAQCAKBIBAIAoEgEAgCgXCJdxQeccXB5SgrDNS+u9OzsINAEAgEgUAQCASBQBAIBIFAEAiEUweFdzXz6tFRnzNzzaO+1wpXvNpBIAgEgkAgCASCQCAIBIJAIAgEgkHhSa54gm/UwHG1z/kJOwgEgUAQCASBQBAIBIFAEAgEgUA4dVB4pysov7vrEHDU54xa8wr/f+wgEAQCQSAQBAJBIBAEAkEgEAQC4elB4cxh2V2tNgQcZYWTgN+5ehROIhAIAoEgEAgCgSAQCAKBIBAIH9u2ba9eBKzKDgJBIBAEAkEgEAQC4S+w0gP+veG1cAAAAABJRU5ErkJggg==" width="200" height="200">
        <div class="qr-content">qrContent</div>
      </div>


    </div>
  </body>

</html>
`;

exports[`@/mixins/print $_printQRs prints an error if no items to print are given 1`] = `
<html>

  <head>
    <style>
      @page {
        size: A4;
        margin: 16px;
      }

      body {
        font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Ubuntu, Helvetica Neue, Arial, sans-serif;
      }

      .part {
        border: 1px solid #aaa;
        width: 4cm;
        float: left;
        padding: 8px;
        margin: 4px;
      }

      .qr-code {
        width: 100% !important;
        height: auto !important;
      }

      .qr-content {
        font-size: 0.7em;
        opacity: 0.7;
        width: 100%;
        word-break:
          break-all;
        word-break: break-word;
        margin: 2px 0 0;
      }

      .title {
        margin: 4px 0;
        font-size: 1.25em;
        font-weight: bold;
      }

      .description {
        font-size: 1em;
        color: #555;
        margin: 0;
        word-break: break-all;
        word-break: break-word;
      }
    </style>
  </head>

  <body>Keine Einträge zu drucken.</body>

</html>
`;
